### ISSUE TITLE
Guide for compiling VSIX Project using Visual Studio 2013

### ISSUE THREAD

**Issue by wycoconut** [readium/readium-sdk#168]
Hi I'm not familiar with GitHub just yet but I'd like to share something for starters.

Would like to provide some detailed steps for compiling the VSIX installer since there is no pre-built version in the official GitHub Repo.  At the end of this article you will see a link to the readium-WindowsRTLauncher by another contributor (@TagCreativeStudio) which maybe useful for someone out there.

Note : Works for Visual Studio 2013 (non express edition)

# Visual Studio Prerequisites
1. Install the **Microsoft Foundation Classes for C++** package for your Visual Studio 2013 installation.  You can do this using the Program and Features --> Modify --> under "Optional Features to install",  check on the **Microsoft Foundation Classes for C++** package.
2. Restart your computer.
3. Install the [Visual Studio 2013 SDK] (http://www.microsoft.com/en-my/download/details.aspx?id=40758).

# Download Readium SDK Solution
1. Download the Readium SDK is downloaded from the official [GitHub Repo (https://github.com/readium/readium-sdk).
2. Open the readium-sdk/Platform/WinRT/Readium/Readium.sln.  You will see two projects : **Readium** and **VSIXReadium**

# Compiling Readium SDK Projects
1. If you try to build the **VSIXReadium** project, it will fail.  ("Path not found", "File not found" errors)
2. To resolve the compilation issue in Step 1, you will need to build the **Readium** project.
3. If you try to build the **Readium** project, it will fail.  ("ambiguous call to overloaded function" errors)
4. Locate the lines with errors and modify them as follows:

```c#
/// run_loop_windows.cpp
ThreadEmulation::TlsFree(RunLoop::RunLoopTLSKey);
RunLoopTLSKey = ThreadEmulation::TlsAlloc();

/// run_loop_common.ipp
# define TLS_GET(key)       ThreadEmulation::TlsGetValue(key)
# define TLS_SET(key, data) ThreadEmulation::TlsSetValue(key, data)
```
5. After applying the fix in Step 4, compile the **Readium** project again and it should pass.
6. Change the Build Configurations and build the **Readium** project.  Repeat for Release-ARM, Release-Win32, Release-x64, Debug-ARM, Debug-Win32, Debug-x64.
![image](https://cloud.githubusercontent.com/assets/5443702/6957954/f2a01a52-d936-11e4-8a56-fadd4c7ec35e.png)
7. When you are done, you should be able to see the output folders in the "deploy" folder.
![image](https://cloud.githubusercontent.com/assets/5443702/6957973/658cfe22-d937-11e4-8b91-fbcd28c5d990.png)
8. Build the **VSIXReadium** project again and it should pass.
9. Do a release build of the **VSIXReadium** project, you should see the output file `VSIXReadium.vsix` in the `bin\Release` folder.

# Installing the VSIX
1. Make sure that Visual Studio is not running.
2. Double-click the `VSIXReadium.vsix` file.  Wait for a while as it gets installed.
3. To verify that it has been installed, start up Visual Studio, goto Tools --> Extensions and Updates.
4. Scroll down the list and you should see VSIXReadium in the list.
![image](https://cloud.githubusercontent.com/assets/5443702/6958122/a7a08f34-d939-11e4-8f39-47be65c44186.png)

After completing all the steps above, you can reference the Readium SDK properly in the [Readium Test App for Windows 8.1 project] (https://github.com/TagCreativeStudio/readium-WindowsRTLauncher#readium-test-app-for-windows-81).

**Comment by danielweck** [comment]
Note that this guide is also available (and maybe updated) at @TagCreativeStudio's own GitHub repository:
https://github.com/TagCreativeStudio/readium-WindowsRTLauncher/issues/1#issue-65840393

**Comment by naveen1941** [comment]
can't build successfully. Lot of errors are 
![capture](https://cloud.githubusercontent.com/assets/4424952/10163655/0452f4c6-66d2-11e5-8e1a-2fa0511d8d5b.PNG)

**Comment by zxz1989** [comment]
can't build successfully. Lot of errors are
![qq 20151020161157](https://cloud.githubusercontent.com/assets/13619754/10601620/6580842e-7745-11e5-9f99-3327d6835f97.png)
how to correct these questions？

**Comment by danielweck** [comment]
@zxz1989 and @naveen1941 the "Windows RunTime" codebase has been contributed by third-parties, and I am afraid ; as far as I know ; that none of the currently active developers in the Readium open-source project(s) have been maintaining the "managed" wrapper code, to remain in sync with the native C++ Readium SDK APIs. Just like the MFC application launcher, there is a regrettable lack of interest from open-source participants for the Windows platform(s).

In fact, the desktop OSX application launcher is not seeing much interest lately. Most vendors of ebook reading systems seem to focus on iOS and Android mobile devices, and also cloud reader products for desktop environments (ReadiumJS for pure web browser runtime, Chrome OS or browser extensions, and additionally Firefox OS). There is also some interest for hybrid apps that use a native shell (user interface + system-level integration) wrapped around a ReadiumJS core (such as Apache Cordova, or other Node + WebKit frameworks).

@rkwright is that a fair summary of the current situation?

**Comment by rkwright** [comment]
That is correct.  We are also looking at another approach, using the Chrome extension with a native part based on PNACL or similar.

**Comment by naveen1941** [comment]
@danielweck I am interested in development for Windows Platform. Could you please guide?

**Comment by pitthaya** [comment]
I want readium-sdk for reading epub at support windows store app.
please update sdk in visual studio 2015.

**Comment by rkwright** [comment]
As an open source project, we are dependent on volunteers to provide support for various platforms.  At present, we support iOS, Android, and OSX.  Unfortunately, nobody has shown any interest in porting Readium to Windows.   One group started a project and I myself ported it to MFC but didn't have time to fully complete the work.  Even Microsoft themselves said they might do it, but they didn't follow through. If you would like to do it, the Readium community would be happy to give you advice as needed.

More info [here](http://readium.github.io/index.html)

Ric Wright
Director of Engineering
Readium Foundation

**Comment by naveen1941** [comment]
Yes @rkwright  I would like to start :)

**Comment by naveen1941** [comment]
I have some progress. I tried readium-js-viewer to show in WPF application using Chromium browser(CEF), and one local host server to serve files required be Chromium. It works fine but its not pretty much fast. I would like to contribute for native readium-js library development.
